Kafka Connect Exporter

changchanghwang opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
Collect connector and task status information from Kafka Connect clusters and exposes them as Prometheus metrics.
- **Key Features:**
  - Collects metrics:
    - total connector count
    - running tasks 
    - failed tasks
    - paused tasks 
    - unassigned tasks 
    - total task count
  - Supports multi-host configurations, allowing for parallel metric collection from multiple Kafka Connect instances with goroutine.
  - Implements robust error handling with detailed logging and error reporting for HTTP errors and JSON parsing issues.

<!--Describe what testing was performed and which tests were added.-->
#### Testing
- Add Unit Tests (average coverage: 62.8%)

<!--Describe the documentation added.-->
#### Documentation

- A detailed `README.md` has been created, covering the project overview, usage instructions, and deployment methods (Docker, Docker Compose, and Kubernetes).
- Additional documentation for contributors has been provided, including contribution guidelines and a code of conduct.

changchanghwang committed:
feat: kafka connect exporter

changchanghwang committed:
docs: add templates and contributing guide, update usage, details

add issue, pull reequest templates
add contributing guide
add code of conduct
update guide for usage
update exported metrics detail

changchanghwang committed:
chore: auto assign

changchanghwang committed:
test: application error package test

changchanghwang committed:
refactor: unexport connector retrieval functions

changchanghwang committed:
test: add collector tests

test: add getConnetorStatus tests

test: add collect tests

changchanghwang committed:
chore: disable cgo, use specific go version

chore: specify image fetch version

titaniper commented on internal/config/config.go:
```diff
@@ -0,0 +1,19 @@
+
+var (
+	Port              = getEnvWithDefault("PORT", "9113")
+	PullingEndpoint   = getEnvWithDefault("PULLING_ENDPOINT", "/metrics")
```
exporter 기준으로 네이밍하면 METRICS_ENDPOINT, METRICS_EXPORT_PATH 등이 더 적절할 수 있을 거 같은데 의견이 궁금합니다.

changchanghwang replied:
METRICS_ENDPOINT 로 변경했습니다 :)

titaniper commented on go.mod:
```diff
@@ -0,0 +1,27 @@
+module github.com/Ecube-labs/kafka-connect-exporter
+
+go 1.23.4
```
Dockerfile 버전이랑 다른데 상관 없을까요?

changchanghwang replied:
dockerfile에서 1.23.4-alpine으로 변경했습니다!

titaniper commented on internal/collector/types.go:
```diff
@@ -0,0 +1,26 @@
+	Tasks []taskStatus `json:"tasks"`
+}
+
+type taskStatus struct {
```
중한 건 아니지만 connectorTaskStatus?

changchanghwang replied:
connector prefix 붙는게 통일성있겠네요. 수정했습니다.

titaniper commented on cmd/exporter/main.go:
```diff
@@ -0,0 +1,25 @@
+func main() {
+	prometheus.MustRegister(collector.New())
+	http.Handle(config.PullingEndpoint, promhttp.Handler())
+	http.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
```
health check endpoint 도 커스텀 가능하게 바꾸는 건 어떨까요?

changchanghwang replied:
그부분 고민하고있었는데 역시 커스텀하게 하는게 좋을 것 같네요 ㅎㅎ 수정했습니다.

titaniper commented on internal/collector/collector.go:
```diff
@@ -0,0 +1,198 @@
+	var wg sync.WaitGroup
+
+	// 각 호스트마다 병렬로 데이터 수집
+	for _, host := range config.KafkaConnectHosts {
```
혹시 과거에 쌓인 메트릭들 reset 되나요?

titaniper replied:
> 혹시 과거에 쌓인 메트릭들 reset 되나요?

@changchanghwang : reset 된다

changchanghwang replied:
네 매트릭을 새로 만들기때문에 기존에 있다가 없어지면 없어집니다.

titaniper commented on README.md:
```diff
@@ -1 +1,115 @@
+
+Exposes metrics based on the collected data:
+
+- `kafka_connect_total`: Total number of connectors.
```
connector_ 들이 누락된 거 아닌지 확인 부탁드립니다.

changchanghwang replied:
누락된거 맞습니다. 수정했습니다. 확인 감사합니다 :)

changchanghwang committed:
refactor: naming, guage description, prefix, translate comment

selenehyun commented on README.md:
```diff
@@ -1 +1,115 @@
+Exposes metrics based on the collected data:
+
+- `kafka_connect_total`: Total number of connectors.
+- `kafka_connect_running`: Number of tasks in the RUNNING state.
+- `kafka_connect_failed`: Number of tasks in the FAILED state.
+- `kafka_connect_paused`: Number of tasks in the PAUSED state.
+- `kafka_connect_unassigned`: Number of tasks with an undefined state (UNKNOWN).
```
https://training.promlabs.com/training/recording-rules/recording-rules-overview/rule-naming-conventions/

running/failed/paused/unassigned 는 상태를 의미해서 어떤 유형(gauge/counter/...)의 지표인지, 지표가 어떤 값을 의미하는지 파악이 어렵습니다

changchanghwang replied:
링크 감사합니다! `_total`을 postfix로 붙혔습니다.

selenehyun commented on README.md:
```diff
@@ -1 +1,115 @@
+- `kafka_connect_running`: Number of tasks in the RUNNING state.
+- `kafka_connect_failed`: Number of tasks in the FAILED state.
+- `kafka_connect_paused`: Number of tasks in the PAUSED state.
+- `kafka_connect_unassigned`: Number of tasks with an undefined state (UNKNOWN).
```
kafka_connect_unassigned는 unknown이 아니고 unassigned 인 이유가 있나요?

changchanghwang replied:
GPT translate error..... 실제로 connect에서 task가 아직 worker에 배정되지 않은경우 UNASSIGNED로 주고있는걸로 알고있습니다.

https://docs.confluent.io/platform/current/connect/monitoring.html#connector-and-task-status

selenehyun commented on README.md:
```diff
@@ -1 +1,115 @@
+- `kafka_connect_paused`: Number of tasks in the PAUSED state.
+- `kafka_connect_unassigned`: Number of tasks with an undefined state (UNKNOWN).
+- `kafka_connect_task_total`: Total number of tasks.
+- `Multi-Host Support`: Configure multiple Kafka Connect instances for parallel metric collection.
+- `Robust Error Handling`: Provides proper logging and error reporting for HTTP errors and JSON parsing issues.
```
이 둘은 지표인건가요?

changchanghwang replied:
기능 설명인데 여기있는게 이상하네요.. 분리했습니다.

selenehyun commented on README.md:
```diff
@@ -1 +1,115 @@
+
+## Features
+
+### Prometheus Metrics Exposure
+
+Exposes metrics based on the collected data:
```
내보내는 지표가 connect의 상태인건가요? connector의 상태인건가요?

그리고 가능하면 레이블을 포함하여 실제로 어떤 형식으로 지표를 내보내고 있는건지 작성되면 좋겠습니다.

changchanghwang replied:
기능에 대한 설명이 가장 초반에 작성해놓고 건들지를 않아서 조금 명확하지 않았던 것 같습니다. 다시 디테일하게 수정했습니다.

changchanghwang committed:
fix: health check endpoint can be provided by environment variables

changchanghwang committed:
test: add GetEnvWithDefault tests

titaniper commented:
근데 이건 배포 어떻게 하나요?, 관련 문서 있는지 궁금

changchanghwang commented:
> 근데 이건 배포 어떻게 하나요?

제가 생각하는 flow는 이렇습니다.
1. pr review
2. pr merge ( + commit message 정리? )
3. release note 작성
4. main 형상 기준으로 docker image build 후 push

> 관련 문서 있는지 궁금

현재는 없습니다. 오픈소스로 운영할 계획이라 어딘가에 작성해야될 것 같은데 어디에 할지 고민되네요... (markdown file? or github  wiki?) 아니면 배포 자체는 저희 회사에서 하는걸로 해서 내부 문서로 남겨놔도 될 것 같구요

selenehyun commented:
가능하면 이런식으로 자동화를 해놔야겠죠

```
name: Build and Push Docker Image

on:
  push:
    branches:
      - main
      - develop

jobs:
  build-and-push:
    runs-on: ubuntu-latest

    steps:
      - name: Check out code
        uses: actions/checkout@v4

      - name: Log in to DockerHub
        uses: docker/login-action@v2
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_PASSWORD }}

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v3

      - name: Determine Docker tag
        id: docker_tag
        run: |
          if [ "${GITHUB_REF_NAME}" == "main" ]; then
            echo "tag=latest" >> $GITHUB_OUTPUT
          elif [ "${GITHUB_REF_NAME}" == "develop" ]; then
            echo "tag=develop" >> $GITHUB_OUTPUT
          else
            echo "tag=${GITHUB_REF_NAME}" >> $GITHUB_OUTPUT
          fi

      - name: Build and push Docker image
        uses: docker/build-push-action@v6
        with:
          context: .
          file: ./Dockerfile
          push: true
          tags: |
            ecubelabs/kafka-connect-exporter:${{ steps.docker_tag.outputs.tag }}
```

changchanghwang committed:
chore: add github action for ci/cd

ci - build, test on pull request, pushed to main
cd - publish docker image based on tag

chore: publish docker image based on tag

changchanghwang commented:
> 가능하면 이런식으로 자동화를 해놔야겠죠

음... 지금은 일단 수동으로 하고 추후에 자동화를 하는건 어떨까요?

자동화를 하더라도 main, develop에 push 시 docker image로 바로 하는것보단 
github tag기반으로 하던가 workflow_dispatch로 수동으로 하는게 좋지 않을까 생각중입니다.

selenehyun commented:
지금은 트리거가 수동이든 자동이든 상관 없고 배포하는 절차는 누구나 똑같이 할수 있어야 하지 않을까요
그래서 그냥 actions 스크립트 작성해두면 될거같은데 문제 있나요?

changchanghwang commented:
> 지금은 트리거가 수동이든 자동이든 상관 없고 배포하는 절차는 누구나 똑같이 할수 있어야 하지 않을까요 그래서 그냥 actions 스크립트 작성해두면 될거같은데 문제 있나요?

아직 어떤것을 기반으로 배포 할 지 정하지 않았어서 추후에 정해진 후 하려고했는데 생각해보니 그냥 지금은 일단 tag 기반으로 한다고 정해놓고 추후에 수정해도 되겠네요

tag 기반으로 배포될 수 있도록 github action workflow 설정했습니다.

titaniper commented:
이건 혹시 결론이 났나요?

selenehyun commented:
0.0.1-beta.0 태그 푸시하면 잘 배포 되는지 확인 부탁드립니다

changchanghwang committed:
chore: update docker hub overview

changchanghwang commented:
> 0.0.1-beta.0 태그 푸시하면 잘 배포 되는지 확인 부탁드립니다

![image](https://github.com/user-attachments/assets/2bbfb9c1-27c4-49d0-90ca-7447b920e208)

잘 되는 것 확인했습니다.

selenehyun commented on .github/workflows/publish-docker-image.yml:
```diff
@@ -0,0 +1,39 @@
+      - name: Log in to DockerHub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
```
fyi @ildukm 

![2025-02-25 at 11 25 14@2x](https://github.com/user-attachments/assets/eba5c796-7238-4b61-b659-0bdc920ed178)
